torch/_subclasses/fake_tensor.py

```diff
@@ -1414,6 +1524,215 @@ def __exit__(self, a, b, c):
             if maybe_prev_fake_mode is not None:
                 torch._C._set_dispatch_mode(maybe_prev_fake_mode)
 
+    @classmethod
+    def cache_info(cls) -> DispatchCacheInfo:
+        """
+        Query the state of the dispatch cache.
+        """
+        return DispatchCacheInfo(
+            FakeTensorMode.cache_hits,
+            FakeTensorMode.cache_misses,
+            dict(FakeTensorMode.cache_bypasses),
+            len(FakeTensorMode.cache),
+        )
+
+    @classmethod
+    def cache_clear(cls):
+        """
+        Clear the dispatch cache.
+        """
+        cls.hits = 0
+        cls.misses = 0
+        cls.cache_bypasses.clear()
+        cls.cache.clear()
+
+    def _cached_dispatch_impl(
+        self,
+        func: OpOverload,
+        types: Tuple[Any],
+        args: Tuple[Any],
+        kwargs: Dict[str, Any],
+    ):
+        """
+        Lookup a cache entry for the given arguments. If none exists, dispatch
+        and cache the result (if the result is cache-able).
+        """
+        try:
+            key = self._cache_key(func, args, kwargs)
+        except _BypassDispatchCache as e:
+            FakeTensorMode.cache_bypasses[e.reason] += 1
+            return self._dispatch_impl(func, types, args, kwargs)
+
+        entry = FakeTensorMode.cache.get(key, None)
+        if entry is not None:
+            FakeTensorMode.cache_hits += 1
+            output = self._output_from_cache_entry(entry, func, args)
+        else:
+            FakeTensorMode.cache_misses += 1
+            output = self._dispatch_impl(func, types, args, kwargs)
+            try:
+                self._create_cache_entry(key, args, output)
+            except _BypassDispatchCache as e:
+                FakeTensorMode.cache_bypasses[e.reason] += 1
+
+        return output
+
+    def _cache_key(
+        self, func: OpOverload, args: Tuple[Any], kwargs: Dict[str, Any]
+    ) -> _DispatchCacheKey:
+        """
+        Create a cache key given the dispatch args. Raise _BypassDispatchCache
+        for any situation that precludes caching.
+        """
+        # Avoid caching for any ops that would require a more sophisticated
+        # caching implementation, e.g., data dependent ops or ops that modify
+        # the inputs.
+        if torch.Tag.data_dependent_output in func.tags:
+            raise _BypassDispatchCache("data dependent output")
+
+        if torch.Tag.dynamic_output_shape in func.tags:
+            raise _BypassDispatchCache("dynamic output shape")
+
+        if torch.Tag.inplace_view in func.tags:
+            raise _BypassDispatchCache("inplace view")
+
+        if func == aten._unsafe_view.default:
+            raise _BypassDispatchCache("unsafe view")
+
+        if func == aten.set_.source_Tensor:
+            raise _BypassDispatchCache("set source")
```

Comment:
Can we add `set_` to `inplace_view` ? cc @bdhirsh

Reply:
yep, I think we should

Reply:
Cool. @@bdhirsh can I get a hint on how to do that. I thought it might be as simple as adding `tags: inplace_view` in native_functions.yaml, but I get an assertion that makes it look like that's not so simple:

`AssertionError: set_.source_Tensor is marked with tag: inplace_view. The codegen expects there to be a corresponding out-of-place view op with the name 'set_' and matching schema, but it didn't find one.`

Reply:
ugh. you're gonna have to massage the codegen a bit, since the codegen makes some assumptions about ops with the `inplace_view` tag, but `set_()` is special. That comment sums it up: codegen expects that every inplace_view has a corresponding out-of-place view op. But `set_()` is a crazy op that swaps out storage, and doesn't have one.

Reply:
It might be enough to add `and str(f.func.name.name) != "set_"` to here https://github.com/pytorch/pytorch/blob/main/torchgen/gen.py#L276. Can you give that a try and see if anything else breaks?

Reply:
I'll remove this once we get https://github.com/pytorch/pytorch/pull/115769 working